parquet-column/src/main/java/org/apache/parquet/column/values/bloom/Bloom.java

```diff
@@ -0,0 +1,485 @@
+
+/**
+ * Bloom Filter is a compact structure to indicate whether an item is not in set or probably
+ * in set. Bloom class is underlying class of Bloom Filter which stores a bit set represents
+ * elements set, hash strategy and bloom filter algorithm.
+ *
+ * Bloom Filter algorithm is implemented using block Bloom filters from Putze et al.'s "Cache-,
+ * Hash- and Space-Efficient Bloom Filters". The basic idea is to hash the item to a tiny Bloom
+ * Filter which size fit a single cache line or smaller. This implementation sets 8 bits in
+ * each tiny Bloom Filter. Tiny bloom filter are 32 bytes to take advantage of 32-bytes SIMD
+ * instruction.
+ */
+
+public abstract class Bloom<T extends Comparable<T>> {
+  // Hash strategy available for bloom filter.
+  public enum HASH {
+    MURMUR3_X64_128,
+  }
+
+  // Bloom filter algorithm.
+  public enum ALGORITHM {
+    BLOCK,
+  }
+
+  /**
+   * Default false positive probability value use to calculate optimal number of bits
+   * used by bloom filter.
+   */
+  public final double FPP = 0.01;
+
+  // Bloom filter data header, including number of bytes, hash strategy and algorithm.
+  public static final int BLOOM_HEADER_SIZE = 12;
+
+  // Bytes in a bucket.
+  public static final int BYTES_PER_BUCKET = 32;
+
+  // Minimum bloom filter data size in byte.
+  public static final int MINIMUM_BLOOM_SIZE = 256;
+
+  // Hash strategy used in this bloom filter.
+  public HASH bloomFilterHash = HASH.MURMUR3_X64_128;
+
+  // Algorithm applied of this bloom filter.
+  public ALGORITHM bloomFilterAlgorithm = ALGORITHM.BLOCK;
+
+  // A hash set to cache the distinct value from column.
+  private HashSet<Long> elements = new HashSet<Long>();
+
+  // The underlying byte array for bloom filter bitset.
+  private byte[] bitset;
+
+  // The size of bitset in byte.
+  private int numBytes;
+
+  // List of byte input to construct the bloom filter.
+  private final List<BytesInput> inputs= new ArrayList<>(4);
+
+  /**
+   * Constructor of bloom filter, if numBytes is zero, bloom filter bitset
+   * will be created lazily and the number of bytes will be calculated through
+   * distinct values in cache.
+   * @param numBytes The number of bytes for bloom filter bitset, set to zero can
+   *                 let it calculate number automatically by using default FPP.
+   * @param hash The hash strategy bloom filter apply.
+   * @param algorithm The algorithm of bloom filter.
+   */
+  public Bloom(int numBytes, HASH hash, ALGORITHM algorithm) {
+    if (numBytes != 0) {
+      initBitset(numBytes);
+    }
+    this.bloomFilterHash = hash;
+    this.bloomFilterAlgorithm = algorithm;
+  }
+
+  /**
+   * Construct the bloom filter with given bit set.
+   * @param bitset The given bitset to construct bloom filter.
+   * @param hash The hash strategy bloom filter apply.
+   * @param algorithm The algorithm of bloom filter.
+   */
+  public Bloom(byte[] bitset, HASH hash, ALGORITHM algorithm) {
```

Comment:
Can you audit which methods are used? I don't see any calls to this.

Reply:
This can be used to construct bloom filter when reading parquet file.

Reply:
That makes sense, but why isn't it then used somewhere else in this patch? Does this patch not exercise the code path of reading a previously-serialized BF? It probably should.

Reply:
Yes, it only contains the Bloom class, Murmur3 class and a unit test for bloom filter functionality.

Both read and write code patch is not in this patch.  I can add some read/write code path unit tests to cover this.